Sources/PackageRegistry/ChecksumTOFU.swift

```diff
@@ -0,0 +1,217 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Dispatch
+
+import Basics
+import PackageFingerprint
+import PackageModel
+
+import struct TSCUtility.Version
+
+struct PackageVersionChecksumTOFU {
+    private let fingerprintStorage: PackageFingerprintStorage?
+    private let fingerprintCheckingMode: FingerprintCheckingMode
+
+    private let registryClient: RegistryClient
+
+    init(
+        fingerprintStorage: PackageFingerprintStorage?,
+        fingerprintCheckingMode: FingerprintCheckingMode,
+        registryClient: RegistryClient
+    ) {
+        self.fingerprintStorage = fingerprintStorage
+        self.fingerprintCheckingMode = fingerprintCheckingMode
+        self.registryClient = registryClient
+    }
+
+    func check(
```

Comment:
nit: `validate` probable a better fit?

Reply:
I will rename it as part of https://github.com/apple/swift-package-manager/pull/6184